File: compiler/nodes/expressions/tuple.go
package expressions

import (
	"wipple/codegen"
	"wipple/database"
	"wipple/syntax"
	"wipple/typecheck"
	"wipple/visit"
)

type TupleExpressionNode struct {
	Elements []database.Node
	Facts    *database.Facts
}

func (node *TupleExpressionNode) GetFacts() *database.Facts {
	return node.Facts
}

func ParseTupleExpression(parser *syntax.Parser) (*TupleExpressionNode, *syntax.Error) {
	span := parser.Spanned()

	many, err := syntax.ParseMany(parser, 1, ParseExpressionElement, func(parser *syntax.Parser) (*struct{}, *syntax.Error) {
		_, err := parser.Token("TupleOperator")
		if err != nil {
			return nil, err
		}

		parser.ConsumeLineBreaks()

		return &struct{}{}, nil
	})
	if err != nil {
		return nil, err
	}

	if len(many) == 1 {
		_, err = parser.Token("TupleOperator")
		if err != nil {
			return nil, err
		}
	} else {
		_, _, err := syntax.ParseOptional(parser, func(parser *syntax.Parser) (struct{}, *syntax.Error) {
			_, err := parser.Token("TupleOperator")
			if err != nil {
				return struct{}{}, err
			}

			return struct{}{}, nil
		})
		if err != nil {
			return nil, err
		}
	}

	elements := make([]database.Node, 0, len(many))
	for _, item := range many {
		elements = append(elements, item.Value)
	}

	return &TupleExpressionNode{
		Elements: elements,
		Facts:    database.NewFacts(span()),
	}, nil
}

func (node *TupleExpressionNode) Visit(visitor *visit.Visitor) {
	visitExpression(visitor, node)

	for _, element := range node.Elements {
		visitor.Visit(element)
	}

	visitor.Constraint(typecheck.NewTypeConstraint(node, typecheck.TupleType(node.Elements)))
}

func (node *TupleExpressionNode) Codegen(c *codegen.Codegen) error {
	span := database.GetSpanFact(node)

	c.WriteString(span, "[")
	for _, element := range node.Elements {
		if err := c.Write(element); err != nil {
			return err
		}

		c.WriteString(span, ", ")
	}
	c.WriteString(span, "]")

	return nil
}

File: compiler/src/nodes/expressions/tuple.rs
use crate::{
    codegen::{Codegen, CodegenCtx, CodegenError},
    database::{Node, NodeRef},
    nodes::{parse_expression_element, visit_expression},
    syntax::{ParseError, Parser, TokenKind},
    typecheck::TypeConstraint,
    visit::{Visit, Visitor},
};

#[derive(Debug)]
pub struct TupleExpressionNode {
    pub elements: Vec<NodeRef>,
}

impl Node for TupleExpressionNode {}

pub fn parse_tuple_expression(parser: &mut Parser<'_>) -> Result<TupleExpressionNode, ParseError> {
    let elements = parser
        .parse_many(1, parse_expression_element, |parser| {
            parser.token(TokenKind::TupleOperator)?;
            parser.consume_line_breaks();
            Ok(())
        })?
        .into_iter()
        .map(|(node, _)| node)
        .collect::<Vec<_>>();

    if elements.len() == 1 {
        parser.token(TokenKind::TupleOperator)?;
    } else {
        let _ = parser.parse_optional(|parser| {
            parser.token(TokenKind::TupleOperator)?;
            Ok(())
        })?;
    }

    Ok(TupleExpressionNode { elements })
}

impl Visit for TupleExpressionNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visit_expression(node, visitor);

        for element in &self.elements {
            visitor.visit(element);
            visitor.edge(element, node, "element");
        }

        visitor.constraint(TypeConstraint::new(
            node.clone(),
            visitor.tuple_type(self.elements.iter().cloned()),
        ));
    }
}

impl Codegen for TupleExpressionNode {
    fn codegen(&self, ctx: &mut CodegenCtx<'_>) -> Result<(), CodegenError> {
        ctx.write_string("[");
        for element in &self.elements {
            ctx.write(element)?;
            ctx.write_string(", ");
        }
        ctx.write_string("]");

        Ok(())
    }
}
